firmware/esp32_ttgo/main.cpp

```diff
@@ -0,0 +1,426 @@
+#include <Arduino.h>
+
+// required for LoRa and/or AsyncSDServer libraries?
+#include <FS.h>
+#include <SPI.h>
+
+#include <WiFi.h>
+#include <AsyncTCP.h>
+#include <ESPAsyncWebServer.h>
+#include <ESPmDNS.h>
+#include <AsyncSDServer.h>
+#include <SD.h>
+#include <SPIFFS.h>
+#include "SSD1306Wire.h"
+
+#include "config.h"
+
+// server
+#include "server/DisasterRadio.h"
+// client
+#include "client/OLEDClient.h"
+#include "client/HistoryRecord.h"
+#include "client/LoRaClient.h"
+#include "client/StreamClient.h"
+#include "client/TCPClient.h"
+#include "client/WebSocketClient.h"
+#include "client/GPSClient.h"
+// middleware
+#include "middleware/Console.h"
+#include "middleware/HistoryReplay.h"
+// history
+#include "history/HistorySD.h"
+#include "history/HistoryMemory.h"
+
+#include "DisasterHistory.h"
+#include "DisasterMiddleware.h"
+
+#define MDNS_NAME_PREFIX "disaster"
+#define HOST_NAME "disaster.radio"
+#define WIFI_AP_SSID_PREFIX "dstr "
+
+// #define WIFI_SSID "WIFI_SSID"
+// #define WIFI_PASSWORD "WIFI_PASSWORD"
+
+#define INDEX_FILE "index.htm"
+
+SSD1306Wire display(0x3c, OLED_SDA, OLED_SCL);
+
+SPIClass sd_card(HSPI);
+
+AsyncServer tcp_server(23);
+AsyncWebServer http_server(80);
+AsyncWebSocket ws_server("/ws");
+
+DisasterRadio *radio = new DisasterRadio();
+DisasterHistory *history = NULL;
+
+int internal_clients = 0;
+int clients = 0;
+int routes = 0;
+
+#define WIFI_POLL_DELAY 500
+#define WIFI_POLL_TRIES 10
+
+char macaddr[12 + 1] = {'\0'};
+char ssid[100] = {'\0'};
+IPAddress ip;
+
+bool sdInitialized = false;
+bool spiffsInitialized = false;
+bool displayInitialized = false;
+bool loraInitialized = false;
+
+void setupWiFi()
+{
+  Serial.println("* Initializing WiFi...");
+
+  // get the WiFi MAC address
+  uint8_t mac[6];
+  WiFi.macAddress(mac);
+
+  // format WiFi MAC address to string
+  sprintf(macaddr, "%02x%02x%02x%02x%02x%02x", mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+
+#ifdef WIFI_SSID
+  Serial.printf(" --> Connecting to WiFi \"%s\"\n", WIFI_SSID);
+  WiFi.begin(WIFI_SSID, WIFI_PASSWORD);
+  for (int i = 0; i < WIFI_POLL_TRIES && WiFi.status() != WL_CONNECTED; i++)
+  {
+    delay(WIFI_POLL_DELAY);
+  }
+  if (WiFi.status() == WL_CONNECTED)
+  {
+    ip = WiFi.localIP();
+    snprintf(ssid, sizeof(ssid), "%s", WIFI_SSID);
+
+    Serial.print(" --> WiFi connected, IP address: ");
+    Serial.println(ip);
+  }
+  else
+  {
+    Serial.println(" --> Unable to connect to WiFi");
+  }
+#endif
+
+  if (WiFi.status() != WL_CONNECTED)
+  {
+    // format full SSID with MAC address suffix
+    snprintf(ssid, sizeof(ssid), "%s%s", WIFI_AP_SSID_PREFIX, macaddr);
+
+    // start the AP
+    WiFi.mode(WIFI_AP);
+    WiFi.setHostname(HOST_NAME);
+    WiFi.softAP(ssid);
+
+    ip = WiFi.localIP();
```

Comment:
I guess the above comment should be placed down here. Same goes, we should figure out a way to dynamically retrieve the real IP address of the node in AP mode, maybe it could be just be set in `config.h`? Or maybe if we create a `preferences` file of some sort, that could then be reconfigured from the web gui or console?